File: 1-Part/4-chapter/practice/addChildren.js
let root = {
    'data': 'Alice', 'children': [
        {
            'data': 'Bob', 'children': [
                {'data': 'Darya', 'children': []}
            ]
        },
        {
            'data': 'Caroline', 'children': [
                {
                    'data': 'Eve', 'children': [
                        {'data': 'Gonzalo', 'children': []},
                        {'data': 'Hadassah', 'children': []}
                    ]
                },
                {'data': 'Fred', 'children': []}
            ]
        }
    ]
}

function addChildren(node) {
    if (node.children.length > 0) {
        node.children.forEach(child => addChildren(child))
    } else {
        node.children.push({'data': 'NewChild', 'children': []})
    }
}

function printTree(node, level = 0) {
    console.log(' '.repeat(level) + node.data)
    node.children.forEach(child => printTree(child, level + 1))
}

printTree(root)
addChildren(root)
printTree(root)

File: 1-Part/4-chapter/practice/addChildren.py
root = {
    'data': 'A', 'children': [
        {'data': 'B', 'children': [
            {'data': 'D', 'children': []}
        ]},
        {'data': 'C', 'children': [
            {'data': 'E', 'children': [
                {'data': 'G', 'children': []},
                {'data': 'H', 'children': []}
            ]},
            {'data': 'F', 'children': []}
        ]}
    ]
}


def addChildren(node):
    if len(node['children']) > 0:
        for child in node['children']:
            addChildren(child)
    else:
        node['children'].append({'data': 'NewChildren', 'children': []})
    return


def printTree(node, level=0):
    print(' ' * level + node['data'])
    for child in node['children']:
        printTree(child, level + 1)


addChildren(root)
printTree(root)
